### PR TITLE
fix diloco 2

### PR DESCRIPTION
**Pull request by samsja**
# FIX DILOCO LFG

This pr fix the diloco intregration.  The main bug was that model were not initialized the same across diloco rank. We used to fix this by pulling the weight init from hf. We don't do this anymore because we don't use transformer for the modeling.

This pr introduces a common seed. I checked, and the weights are correctly initialized now.
Screenshot from 2024-09-26 18-29-49

The loss is looking better now
Screenshot from 2024-09-26 18-28-01

Additionally the PR add testing for the diloco all reduce as well as some small fix and enhancement

### CODE REVIEW

File: src/zeroband/train.py [hunk 1]
```diff
@@ -124,12 +130,6 @@ def train(config: Config):
         model = torch.compile(model)
     logger.debug("model compiled and fsdped")
 
-    if config.diloco is not None:
-        if world_info.local_world_size == 1:
-            raise ValueError("Diloco is not supported for local_world_size == 1 because of a pytorch bug")
-
-        diloco = Diloco(config.diloco, model, sharding_strategy, elastic_device_mesh)
-
```
Comment (samsja) [inline src/zeroband/train.py:132]: just move after adamw innit

File: src/zeroband/diloco.py [hunk 2]
```diff
@@ -66,11 +66,11 @@ def __init__(
         config: DilocoConfig,
         model: nn.Module,
         fsdp_sharding_strategy: ShardingStrategy,
-        elastic_device_mesh: ElasticDeviceMesh,
+        global_pg: dist.ProcessGroup,
```
Comment (JohannesHa) [inline src/zeroband/diloco.py:69]: why did we switch back from having the elastic device mesh to just one process group?

Comment (samsja) [reply src/zeroband/diloco.py:69]: diloco only needs the global pg. ElasticDeviceMesh exposes two dynamic process groups global and local. 

Might switch it back with @Jackmin801 PR

Comment (Jackmin801) [reply src/zeroband/diloco.py:69]: Hrmm I think my PR doesnt use anything from ElasticDeviceMesh other than the global_pg so should be fine. We might need some inner stuff in the future but we can leave that to the future I suppose.

I use some `dist.barrier()` which in a way uses the inner pg already.

File: src/zeroband/diloco.py [hunk 3]
```diff
@@ -93,22 +93,21 @@ def sync_pseudo_gradient(self, model: nn.Module):
...
     def sync_inner_model(self, model: nn.Module):
         """
-        Sync the inner model from the global process group to the local process group
+        Sync the inner model from the CPU outer model to GPU
         """
 
         self._logger.debug("sync inner model")
         for param_offloaded, param in zip(self.param_list_cpu, model.parameters()):
-            param.data = param_offloaded.data.to("cuda")  # todo: use copy_ here
+            param.data.copy_(param_offloaded.data.to(param.device))  # todo: use copy_ here
```
Comment (Jackmin801) [inline src/zeroband/diloco.py]: the .to(device) isnt needed here. You should be able to directly copy the data

Comment (samsja) [reply src/zeroband/diloco.py]: interesting so it will handle the move to device by itself ?

File: src/zeroband/train.py (hunk 1 above)
Comment (Jackmin801) [reply src/zeroband/train.py:132]: ook

File: src/zeroband/diloco.py (hunk 3 above)
Comment (Jackmin801) [reply src/zeroband/diloco.py]: my understanding is that a.copy_(b) means copy the content of b to a inplace so it works across devices and doesnt allocate a new intermediate storage.

meanwhile param.copy_(param_offloaded.data.to(param.device)) will first evaluate the `.to` which creates a new intermediate storage then copies the content. might be wrong tho, not sure how smart compilers/interpreters are

Comment (samsja) [reply src/zeroband/diloco.py]: done

**Review by Jackmin801 — APPROVED** [review]
LFGTM!
